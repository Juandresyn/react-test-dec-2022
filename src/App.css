#root {
  width: 1100px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
  --nav-bar-width: 230px;
}

.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
}
.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}
.logo.react:hover {
  filter: drop-shadow(0 0 2em #61dafbaa);
}

@media (prefers-reduced-motion: no-preference) {
  a:nth-of-type(2) .logo {
    animation: logo-spin infinite 20s linear;
  }
}

.card {
  padding: 2em;
}

.main {
  display: flex;
}

.toolbar {
  width: var(--nav-bar-width);
  text-align: left;
  padding-right: 30px;
  margin-top: 120px;
}

.menu {
  padding: 0;
  margin: 0;
  list-style: none;
  text-transform: uppercase;
}

.menu li {
  padding: 10px 0;
}

.menu a {
  font-weight: bold;
}

.content {
  width: calc(100% - var(--nav-bar-width));
}