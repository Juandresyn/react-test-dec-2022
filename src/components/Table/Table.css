.table {
    width: 100%;
    border-collapse: collapse;
}

.table th {
    border-bottom: 1px solid #ccc;
    padding: 10px 5px;
}

.table tbody td {
    border: 0;
    border-bottom: 1px solid #ccc;
    padding: 10px 5px;
    text-transform: uppercase;
    font-weight: 500;
}

.table tbody tr:nth-child(odd) {
    background-color: rgba(255, 255, 255, .05)
}