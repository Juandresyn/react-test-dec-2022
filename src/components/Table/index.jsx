/**
* Returns array of values from an object
*/
const deconstructItem = (item) => Object.values(item);

function Table({ headers, items, options }) {
  /**
  * Check if table props contains remove function.
  */
  const hasRemove = !!options && !!options.remove;

  return (
    <table className="table">
      <thead>
        <tr>
          { headers.map((h, index) => <th key={`th_${index}`}>{ h.toUpperCase() }</th>) }
          { !!options ? <th>OPTIONS</th> : null }
        </tr>
      </thead>
      <tbody>
        { items.map((i, index) => {
          return (
            <tr key={`item_${index}`} id={i.id}>
              { deconstructItem(i).map((value, index) => <td key={`td_${index}`}>{ value }</td> ) }
              { hasRemove ? <td><button id={`remove__${ i.id }`} role="button" type="button" onClick={()=> options.remove(i.id)}>Remove</button></td> : null }
            </tr>
          )
        }) }
      </tbody>
    </table>
  )
}

export default Table;
