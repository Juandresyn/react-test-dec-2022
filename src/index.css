:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  display: flex;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.button--wide {
  width: 100%;
  margin-bottom: 30px;
}

.form {
  display: flex;
  flex-direction: column;
  margin-bottom: 35px;
}

.form__wrapper {
  display: flex;
  flex-direction: column;
}

.form__group {
  display: flex;
  justify-content: space-between;
}

.form__group > * {
  width: 49% !important;
}

.form__wrapper input,
.form__wrapper select,
.form__wrapper textarea {
  width: 100%;
  padding: 15px 10px;
  border: 1px solid #cdcdcd;
  border-radius: 3px;
  margin-bottom: 14px;
  box-sizing: border-box;
}

.title__main {
  text-align: left;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}
