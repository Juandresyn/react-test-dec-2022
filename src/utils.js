export const transformDate = (date, showTime = true) => {
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    const hour = date.getHours();
    const minutes = date.getMinutes();

    const time = `${hour}:${minutes}:00`;

    return `${year}-${month}-${day} ${showTime ? time : ''}`
};

export const getAge = (dob) => {
    const ageDifMs = Date.now() - new Date(dob).getTime();
    const ageDate = new Date(ageDifMs);

    return Math.abs(ageDate.getUTCFullYear() - 1970);
}

export const users = [
    {
        id: '23456789',
        name: 'Pepito',
        lastName: 'Perez',
        dob: '1973-06-12'
    },
    {
        id: '12345678',
        name: 'Juan',
        lastName: 'Yepes',
        dob: '1992-02-16'
    }
];

export const cars = [
    {
      id: "NOF123",
      maker: 'mercedes-benz',
      model: 2023,
      ref: 'GLC300',
      color: 'blue',
      milage: 31278
    },
    {
      id: "IHJ889",
      maker: 'audi',
      model: 2020,
      ref: 'Q5',
      color: 'white',
      milage: 45123
    },
    {
      id: "UUE654",
      maker: 'mazda',
      model: 2021,
      ref: 'CX-9',
      color: 'silver',
      milage: 55687
    }
];

export const reservations = [
    {
      id: "rid-20221210-000001",
      user: {
        id: '12345678',
        name: 'Juan',
        lastName: 'Yepes',
        age: '30'
      },
      car: {
        id: "NOF123",
        maker: 'mercedes-benz',
        model: 2023,
        ref: 'GLC300',
        color: 'blue',
        milage: 31278
      },
      from: '2022-12-10 10:00:00',
      to: '2022-12-20 11:00:00',
      notes: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed volutpat, nibh nec egestas sodales, justo nulla sodales dolor.'
    },
    {
      id: "rid-20221210-000002",
      user: {
        id: '23456789',
        name: 'Pepito',
        lastName: 'Perez',
        age: '45'
      },
      car: {
        id: "IHJ889",
        maker: 'audi',
        model: 2020,
        ref: 'Q5',
        color: 'white',
        milage: 45123
      },
      from: '2022-12-10 10:30:00',
      to: '2022-12-20 11:10:00',
      notes: ''
    },
];