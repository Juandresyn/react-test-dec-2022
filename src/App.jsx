import { useState } from 'react';
import {
  createBrowserRouter,
  RouterProvider,
} from "react-router-dom";
import reactLogo from './assets/react.svg';
import Cars from './components/Cars';
import Users from './components/Users';
import Reservations from './components/Reservations';
import Logo from './assets/logo.png';
import './App.css';

const router = createBrowserRouter([
  {
    path: "/",
    element: <h1>Rent-a-Car</h1>,
  },
  {
    path: "/manage/cars",
    element: <Cars></Cars>,
  },
  {
    path: "/manage/users",
    element: <Users></Users>,
  },
  {
    path: "/manage/reservations",
    element: <Reservations></Reservations>,
  },
]);

function App() {
  const [count, setCount] = useState(0)

  return (
    <div className="App">
      <div>
        <a href="/">
          <img src={ Logo } className="logo" alt="Company Logo" />
        </a>
      </div>
      <main className="main">
        <nav className="toolbar">
          <ul className="menu">
            <li>
              <a href="/manage/cars">Manage Cars</a>
            </li>
            <li>
              <a href="/manage/users">Manage Users</a>
            </li>
            <li>
              <a href="/manage/reservations">Manage Reservations</a>
            </li>
          </ul>
        </nav>
        <div className="content">
          <RouterProvider router={router} />
        </div>
      </main>
    </div>
  )
}

export default App;
