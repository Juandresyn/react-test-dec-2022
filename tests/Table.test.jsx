import {
  render, screen, waitFor, act,
} from '@testing-library/react';
import Table from '../src/components/Table';
import { cars } from '../src/mockData';

const headers = ['id', 'license', 'maker', 'model', 'ref', 'color', 'milage'];

describe('Test <Table /> component', () => {
  test('Check fragment', async () => {
    let Render;

    act(() => {
      Render = render(<Table
        headers={headers}
        items={cars}
      />);
    });

    await waitFor(() => {
      const plainHTML = Render.asFragment();
      expect(plainHTML).toMatchSnapshot();
    });
  });

  test('Show render equal number of rows as items passed', async () => {
    act(() => {
      render(<Table
        headers={headers}
        items={cars}
      />);
    });

    await waitFor(() => {
      const license = screen.getByText(`${cars[0].id}`);
      const license2 = screen.getByText(`${cars[1].id}`);
      const license3 = screen.getByText(`${cars[2].id}`);

      expect(license).toBeTruthy();
      expect(license2).toBeTruthy();
      expect(license3).toBeTruthy();
    });
  });

  test('Remove button should display', async () => {
    let Render;

    act(() => {
      Render = render(<Table
        headers={headers}
        items={cars}
        options={({ remove: () => null })}
      />);
    });

    await waitFor(() => {
      const plainHTML = Render.asFragment();
      expect(plainHTML).toMatchSnapshot();

      const removeBtns = Render.container.querySelector(`#remove__${cars[0].id}`);
      expect(removeBtns).toBeTruthy();
    });
  });
});
