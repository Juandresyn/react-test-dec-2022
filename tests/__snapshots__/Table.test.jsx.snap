// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Table /> component Check fragment 1`] = `
<DocumentFragment>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th>
          ID
        </th>
        <th>
          LICENSE
        </th>
        <th>
          MAKER
        </th>
        <th>
          MODEL
        </th>
        <th>
          REF
        </th>
        <th>
          COLOR
        </th>
        <th>
          MILAGE
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        id="NOF123"
      >
        <td>
          NOF123
        </td>
        <td>
          mercedes-benz
        </td>
        <td>
          2023
        </td>
        <td>
          GLC300
        </td>
        <td>
          blue
        </td>
        <td>
          31278
        </td>
      </tr>
      <tr
        id="IHJ889"
      >
        <td>
          IHJ889
        </td>
        <td>
          audi
        </td>
        <td>
          2020
        </td>
        <td>
          Q5
        </td>
        <td>
          white
        </td>
        <td>
          45123
        </td>
      </tr>
      <tr
        id="UUE654"
      >
        <td>
          UUE654
        </td>
        <td>
          mazda
        </td>
        <td>
          2021
        </td>
        <td>
          CX-9
        </td>
        <td>
          silver
        </td>
        <td>
          55687
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Test <Table /> component Remove button should display 1`] = `
<DocumentFragment>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th>
          ID
        </th>
        <th>
          LICENSE
        </th>
        <th>
          MAKER
        </th>
        <th>
          MODEL
        </th>
        <th>
          REF
        </th>
        <th>
          COLOR
        </th>
        <th>
          MILAGE
        </th>
        <th>
          OPTIONS
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        id="NOF123"
      >
        <td>
          NOF123
        </td>
        <td>
          mercedes-benz
        </td>
        <td>
          2023
        </td>
        <td>
          GLC300
        </td>
        <td>
          blue
        </td>
        <td>
          31278
        </td>
        <td>
          <button
            id="remove__NOF123"
            role="button"
          >
            Remove
          </button>
        </td>
      </tr>
      <tr
        id="IHJ889"
      >
        <td>
          IHJ889
        </td>
        <td>
          audi
        </td>
        <td>
          2020
        </td>
        <td>
          Q5
        </td>
        <td>
          white
        </td>
        <td>
          45123
        </td>
        <td>
          <button
            id="remove__IHJ889"
            role="button"
          >
            Remove
          </button>
        </td>
      </tr>
      <tr
        id="UUE654"
      >
        <td>
          UUE654
        </td>
        <td>
          mazda
        </td>
        <td>
          2021
        </td>
        <td>
          CX-9
        </td>
        <td>
          silver
        </td>
        <td>
          55687
        </td>
        <td>
          <button
            id="remove__UUE654"
            role="button"
          >
            Remove
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
